pkg/controller/job/indexed_job_utils.go

```diff
@@ -39,85 +39,184 @@ func isIndexedJob(job *batch.Job) bool {
 	return job.Spec.CompletionMode != nil && *job.Spec.CompletionMode == batch.IndexedCompletion
 }
 
-// calculateSucceededIndexes returns a string representation of the list of
-// succeeded indexes in compressed format and the number of succeeded indexes.
-func calculateSucceededIndexes(pods []*v1.Pod, completions int32) (string, int32) {
-	sort.Sort(byCompletionIndex(pods))
-	var result strings.Builder
-	var lastSucceeded int
-	var count int32
-	firstSucceeded := math.MinInt32
+type interval struct {
+	First int
+	Last  int
+}
+
+type orderedIntervals []interval
+
+// calculateSucceededIndexes returns the old and new list of succeeded indexes
+// in compressed format (intervals).
+// The old list is solely based off .status.completedIndexes, but returns an
+// empty list if this Job is not tracked with finalizers. The new list includes
+// the indexes that succeeded since the last sync.
+func calculateSucceededIndexes(job *batch.Job, pods []*v1.Pod) (orderedIntervals, orderedIntervals) {
```

Comment:
is there a way to isolate this change so the existing logic is used if the feature is disabled? it's hard to reason about the safety of this change for existing jobs when the new feature is off

Reply:
the way would be to revert all the changes and add alternate implementations. I could do that, but we will probably miss the code freeze deadline. I added a lot of coverage for this.